File: Doc-Scanner-X/Doc-Scanner-X/LocalFileTableViewController.m
//
//  LocalFileTableViewController.m
//  Doc-Scanner-X
//
//  Created by dynamsoft on 2017/11/27.
//  Copyright © 2017年 com.dynamsoft. All rights reserved.
//

#import "LocalFileTableViewController.h"

@interface LocalFileTableViewController ()

@end

@implementation LocalFileTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.tableView.separatorInset = UIEdgeInsetsMake(0, 17, 0, 0);
    self.navigationItem.title = @"Saved Files";
    
    self.emptyImage = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"empty"]];
    self.emptyImage.center = CGPointMake([UIScreen mainScreen].bounds.size.width/2, 200);
    self.emptyHint = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 250, 44)];
    self.emptyHint.center = CGPointMake([UIScreen mainScreen].bounds.size.width/2, 350);
    self.emptyHint.textColor = [UIColor grayColor];
    self.emptyHint.textAlignment = NSTextAlignmentCenter;
    self.emptyHint.text = @"You have not saved any files";
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (self.localDatas == nil) {
        return 0;
    }
    return self.localDatas.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell* cell = [[CustomTableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"Cell"];
    if (self.localDatas == nil) {
        return cell;
    }
    cell.textLabel.text = self.localDatas[indexPath.row].dataName;
    cell.textLabel.lineBreakMode = NSLineBreakByTruncatingTail;
    cell.detailTextLabel.text = self.localDatas[indexPath.row].dataTimeStamp;
    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    cell.imageView.contentMode = UIViewContentModeScaleAspectFit;
    NSURL* fileURL = [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask].firstObject URLByAppendingPathComponent:[@"/" stringByAppendingString:self.localDatas[indexPath.row].dataName]];
    cell.imageView.image = [self getThumbnailImageFrom:fileURL Type:self.localDatas[indexPath.row].dataType];
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 94;
}

- (UIView *)tableView:(UITableView *)tableView viewForFooterInSection:(NSInteger)section {
    if (self.localDatas == nil) {
        self.view.backgroundColor = [UIColor colorWithRed:240/255.0 green:239/255.0 blue:244/255.0 alpha:1];
        UIView* uiView = [[UIView alloc] initWithFrame:self.tableView.frame];
        [uiView addSubview:self.emptyHint];
        [uiView addSubview:self.emptyImage];
        return uiView;
    }
    if (self.localDatas.count != 0) {
        self.view.backgroundColor = [UIColor clearColor];
        UIView* uiView = [[UIView alloc] initWithFrame:CGRectZero];
        return uiView;
    } else {
        self.view.backgroundColor = [[UIColor alloc] initWithRed:240/255.0 green:239/255.0 blue:244/255.0 alpha:1];
        UIView* uiView = [[UIView alloc] initWithFrame:self.tableView.frame];
        [uiView addSubview:self.emptyHint];
        [uiView addSubview:self.emptyImage];
        return uiView;
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section {
    return 1;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    if (self.localDatas[indexPath.row] == nil) {
        return;
    }
    self.quickLookVC = [QuickLookViewController new];
    self.quickLookVC.navigationItem.title = @"Preview";
    [self.navigationController pushViewController:self.quickLookVC animated:NO];
    [self.quickLookVC quickLookWithType:self.localDatas[indexPath.row].dataType In:self.localDatas[indexPath.row].dataName];
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}

- (NSArray<UITableViewRowAction *> *)tableView:(UITableView *)tableView editActionsForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewRowAction* deleteAction = [UITableViewRowAction rowActionWithStyle:UITableViewRowActionStyleDestructive title:@"Delete" handler:^(UITableViewRowAction * _Nonnull action, NSIndexPath * _Nonnull indexPath) {
        [self deleteFile:self.localDatas[indexPath.row].dataName];
        [self.localDatas removeObjectAtIndex:indexPath.row];
        [NSKeyedArchiver archiveRootObject:self.localDatas toFile:[LocalData getArchiveURL].path];
        [self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
        if (self.localDatas.count == 0) {
            [self reload];
        }
    }];
    return @[deleteAction];
}

- (void) reload {
    self.localDatas = [[(NSMutableArray<LocalData*>*)[NSKeyedUnarchiver unarchiveObjectWithFile:[LocalData getArchiveURL].path] reverseObjectEnumerator] allObjects];
    [self.tableView reloadData];
}

- (void) deleteFile:(NSString*)fileName {
    NSURL* path = [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask].firstObject URLByAppendingPathComponent:[@"/" stringByAppendingString:fileName]];
    [[NSFileManager defaultManager] removeItemAtURL:path error:nil];
}

- (UIImage*) getThumbnailImageFrom:(NSURL*)fileURL Type:(NSUInteger)type {
    switch (type) {
        case PNG:
        case JPEG:
        {
            _image = [UIImage imageWithContentsOfFile:fileURL.path];
            CGSize imageSize = CGSizeMake(60, 60);
            if (_image.size.width > _image.size.height) {
                imageSize.height = 60 * _image.size.height / _image.size.width;
            } else {
                imageSize.width = 60 * _image.size.width / _image.size.height;
            }
            UIGraphicsBeginImageContextWithOptions(imageSize, NO, [UIScreen mainScreen].scale);
            CGRect imageRect = CGRectMake(0, 0, imageSize.width, imageSize.height);
            [_image drawInRect:imageRect];
            _image = UIGraphicsGetImageFromCurrentImageContext();
            UIGraphicsEndImageContext();
            break;
        }
        case PDF:
        {
            CGPDFDocumentRef pdf = CGPDFDocumentCreateWithURL((CFURLRef)fileURL);
            CGPDFPageRef firstPage = CGPDFDocumentGetPage(pdf, 1);
            CGRect pageRect = CGPDFPageGetBoxRect(firstPage, kCGPDFMediaBox);
            UIGraphicsBeginImageContext(pageRect.size);
            CGContextRef context = UIGraphicsGetCurrentContext();
            CGContextSaveGState(context);
            CGContextTranslateCTM(context, 0, pageRect.size.height);
            CGContextScaleCTM(context, 1, -1);
            CGContextConcatCTM(context, CGPDFPageGetDrawingTransform(firstPage, kCGPDFMediaBox, pageRect, 0, YES));
            CGContextDrawPDFPage(context, firstPage);
            CGContextRestoreGState(context);
            _image = UIGraphicsGetImageFromCurrentImageContext();
            UIGraphicsEndPDFContext();
            CGSize imageSize = CGSizeMake(60, 60);
            if (_image.size.width > _image.size.height) {
                imageSize.height = 60 * _image.size.height / _image.size.width;
            } else {
                imageSize.width = 60 * _image.size.width / _image.size.height;
            }
            UIGraphicsBeginImageContextWithOptions(imageSize, NO, [UIScreen mainScreen].scale);
            CGRect imageRect = CGRectMake(0, 0, imageSize.width, imageSize.height);
            [_image drawInRect:imageRect];
            _image = UIGraphicsGetImageFromCurrentImageContext();
            UIGraphicsEndImageContext();
            break;
        }
        default:
            return nil;
    }
    return _image;
}

@end

File: DocScannerX/DocScannerX/LocalFileTableViewController.swift
//
//  LocalFileTableViewController.swift
//  DCS SuperScanner
//
//  Created by dynamsoft on 2017/11/2.
//  Copyright © 2017年 com.dynamsoft. All rights reserved.
//

import UIKit

class LocalFileTableViewController: UITableViewController {

    // MARK: - Properties
    
    var image: UIImage?
    var quickLookVC: QuickLookViewController?
    var localDatas: [LocalData]?
    var emptyImage: UIImageView?
    var emptyHint: UILabel?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        /// Custom `tableView.separatorInset`.
        tableView.separatorInset = UIEdgeInsets(top: 0, left: 17, bottom: 0, right: 0)
        navigationItem.title = "Saved Files"
        
        /// Initialize `emptyImage` and `emptyHint`.
        emptyImage = UIImageView(image: UIImage(named: "empty"))
        emptyImage!.center = CGPoint(x: fullScreenSize.width/2, y: 200)
        emptyHint = UILabel(frame: CGRect(x: 0, y: 0, width: 250, height: 44))
        emptyHint?.center = CGPoint(x: fullScreenSize.width/2, y: 350)
        emptyHint?.textColor = .gray
        emptyHint?.textAlignment = .center
        emptyHint?.text = "You have not saved any files"
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    
    // MARK: - Table view data source
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        guard let count = localDatas?.count else {
            return 0
        }
        return count
    }

    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = CustomTableViewCell(style: .subtitle, reuseIdentifier: "Cell")
        guard let _ = localDatas else {
            return cell
        }
        
        cell.textLabel?.text = localDatas?[indexPath.row].dataName
        cell.textLabel?.lineBreakMode = .byTruncatingTail
        cell.detailTextLabel?.text = localDatas?[indexPath.row].dataTimeStamp
        cell.accessoryType = .disclosureIndicator
        cell.imageView?.contentMode = .scaleAspectFit
        cell.imageView?.image = getThumbnailImageFrom(FileManager().urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent("/"+localDatas![indexPath.row].dataName), type: localDatas![indexPath.row].dataType)
        return cell
    }
    
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 94
    }
    
    override func tableView(_ tableView: UITableView, viewForFooterInSection section: Int) -> UIView? {
        guard  let _ = localDatas else {
            view.backgroundColor = UIColor(red: 240/255, green: 239/255, blue: 244/255, alpha: 1)
            let uiView = UIView(frame: tableView.frame)
            uiView.addSubview(emptyHint!)
            uiView.addSubview(emptyImage!)
            return uiView
        }
        if localDatas?.count != 0 {
            view.backgroundColor = .clear
            let uiView = UIView(frame: .zero)
            return uiView
        } else {
            view.backgroundColor = UIColor(red: 240/255, green: 239/255, blue: 244/255, alpha: 1)
            let uiView = UIView(frame: tableView.frame)
            uiView.addSubview(emptyHint!)
            uiView.addSubview(emptyImage!)
            return uiView
        }
    }

    override func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        return 1
    }
    
    // MARK: - Table view delegate
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        guard let index = localDatas?[indexPath.row] else {
            return
        }
        quickLookVC = QuickLookViewController()
        quickLookVC?.navigationItem.title = "Preview"
        self.navigationController?.pushViewController(quickLookVC!, animated: false)
        
        quickLookVC?.quickLook(type: index.dataType, in: index.dataName)
    }

    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return true
    }
    
    override func tableView(_ tableView: UITableView, editActionsForRowAt indexPath: IndexPath) -> [UITableViewRowAction]? {
        let deleteAction = UITableViewRowAction(style: .destructive, title: "Delete", handler: {
            (_: UITableViewRowAction, indexPath: IndexPath) in
            self.deleteFile(fileName: (self.localDatas?[indexPath.row].dataName)!)
            self.localDatas?.remove(at: indexPath.row)
            NSKeyedArchiver.archiveRootObject(self.localDatas!, toFile: LocalData.ArchiveURL.path)
            tableView.deleteRows(at: [indexPath], with: .fade)
            /// When delete cells until `localDatas?.count == 0`, reload tableview to show specified footerView.
            if self.localDatas?.count == 0 {
                self.reload()
            }
        })
        return [deleteAction]
    }
    
    /// Retrieve local files information and reload `tableView`.
    func reload() {
        localDatas = (NSKeyedUnarchiver.unarchiveObject(withFile: LocalData.ArchiveURL.path) as? [LocalData])?.reversed()
        tableView.reloadData()
    }
    
    /**
     Delete file saved in APP's sandbox according to file path.
     
     - Parameter fileName: File name
     */
    private func deleteFile(fileName: String) {
        let path = FileManager().urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent("/"+fileName)
        try! FileManager().removeItem(at: path)
    }

    /**
     Generate a thumbnail image from local files. If file type is PNG/JPEG, generate its thumbnail image.
     If file type is PDF, generate its first page's thumbnail image.
     
     - Parameters:
        - path: File url
        - type: File type
     
     - Returns: Thumbnail image
     */
    private func getThumbnailImageFrom(_ path: URL, type: UInt) -> UIImage? {
        switch type {
        case FileType.PNG.rawValue, FileType.JPEG.rawValue:
            image = UIImage(contentsOfFile: path.path)
            var imageSize = CGSize(width: 60, height: 60)
            if image!.size.width > image!.size.height {
                imageSize.height = 60*image!.size.height/image!.size.width
            } else {
                imageSize.width = 60*image!.size.width/image!.size.height
            }
            UIGraphicsBeginImageContextWithOptions(imageSize, false, UIScreen.main.scale)
            let imageRect = CGRect(x: 0, y: 0, width: imageSize.width, height: imageSize.height)
            image?.draw(in: imageRect)
            image = UIGraphicsGetImageFromCurrentImageContext()
            UIGraphicsEndImageContext()
        case FileType.PDF.rawValue:
            let pdf = CGPDFDocument(path as CFURL)
            let firstPage = pdf?.page(at: 1)
            let pageRect = firstPage?.getBoxRect(.mediaBox)
            UIGraphicsBeginImageContext(pageRect!.size)
            let context = UIGraphicsGetCurrentContext()
            context?.saveGState()
            context?.translateBy(x: 0, y: pageRect!.size.height)
            context?.scaleBy(x: 1, y: -1)
            context?.concatenate((firstPage?.getDrawingTransform(.mediaBox, rect: pageRect!, rotate: 0, preserveAspectRatio: true))!)
            context?.drawPDFPage(firstPage!)
            context?.restoreGState()
            image = UIGraphicsGetImageFromCurrentImageContext()!
            UIGraphicsEndImageContext()
            var imageSize = CGSize(width: 60, height: 60)
            if image!.size.width > image!.size.height {
                imageSize.height = 60*image!.size.height/image!.size.width
            } else {
                imageSize.width = 60*image!.size.width/image!.size.height
            }
            UIGraphicsBeginImageContextWithOptions(imageSize, false, UIScreen.main.scale)
            let imageRect = CGRect(x: 0, y: 0, width: imageSize.width, height: imageSize.height)
            image?.draw(in: imageRect)
            image = UIGraphicsGetImageFromCurrentImageContext()
            UIGraphicsEndImageContext()
        default:
            return nil
        }
        return image
    }
    
}
